File: core/src/test/java/uk/lobsterdoodle/namepicker/adapter/AdapterDataWrapperTest.java
package uk.lobsterdoodle.namepicker.adapter;

import org.junit.Before;
import org.junit.Test;

import static java.util.Arrays.asList;
import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.junit.Assert.assertThat;

public class AdapterDataWrapperTest {
    private AdapterDataWrapper<TestClass> wrapper;

    @Before
    public void setUp() {
        wrapper = new AdapterDataWrapper<>();
    }

    @Test
    public void get_count() {
        wrapper.replaceData(asList(testClass(1), testClass(2), testClass(3)));
        assertThat(wrapper.count(), is(equalTo(3)));
    }

    @Test
    public void get_item() {
        wrapper.replaceData(asList(testClass(1), testClass(2), testClass(3)));
        assertThat(wrapper.item(0), is(equalTo(testClass(1))));
        assertThat(wrapper.item(1), is(equalTo(testClass(2))));
        assertThat(wrapper.item(2), is(equalTo(testClass(3))));
    }

    @Test
    public void replace_item() {
        wrapper.replaceData(asList(testClass(1), testClass(2), testClass(3)));
        wrapper.replaceItem(1, testClass(4));
        assertThat(wrapper.item(0), is(equalTo(testClass(1))));
        assertThat(wrapper.item(1), is(equalTo(testClass(4))));
        assertThat(wrapper.item(2), is(equalTo(testClass(3))));
    }

    @Test
    public void replace_data() {
        wrapper.replaceData(asList(testClass(1), testClass(2), testClass(3)));
        wrapper.replaceData(asList(testClass(4), testClass(5), testClass(6)));
        assertThat(wrapper.item(0), is(equalTo(testClass(4))));
        assertThat(wrapper.item(1), is(equalTo(testClass(5))));
        assertThat(wrapper.item(2), is(equalTo(testClass(6))));
    }

    @Test
    public void modify_data() {
        wrapper.replaceData(asList(testClass(1), testClass(2), testClass(3)));
        wrapper.modifyData(input -> new TestClass(input.id + 3));
        assertThat(wrapper.item(0), is(equalTo(testClass(4))));
        assertThat(wrapper.item(1), is(equalTo(testClass(5))));
        assertThat(wrapper.item(2), is(equalTo(testClass(6))));
    }

    private TestClass testClass(int id) {
        return new TestClass(id);
    }

    private class TestClass {
        final int id;

        TestClass(int id) {
            this.id = id;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof TestClass)) return false;

            TestClass testClass = (TestClass) o;

            return id == testClass.id;

        }

        @Override
        public int hashCode() {
            return id;
        }

        @Override
        public String toString() {
            return "TestClass{" +
                    "id=" + id +
                    '}';
        }
    }
}

File: core/src/test/java/uk/lobsterdoodle/namepicker/adapter/AdapterDataWrapperTest.kt
package uk.lobsterdoodle.namepicker.adapter

import org.junit.Before
import org.junit.Test

import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsEqual.equalTo
import org.junit.Assert.assertThat

class AdapterDataWrapperTest {
    private lateinit var wrapper: AdapterDataWrapper<TestClass>

    @Before
    fun setUp() {
        wrapper = AdapterDataWrapper<TestClass>()
    }

    @Test
    fun get_count() {
        wrapper.replaceData(listOf(TestClass(1), TestClass(2), TestClass(3)))
        assertThat(wrapper.count(), `is`(equalTo(3)))
    }

    @Test
    fun get_item() {
        wrapper.replaceData(listOf(TestClass(1), TestClass(2), TestClass(3)))
        assertThat(wrapper.item(0), `is`(equalTo(TestClass(1))))
        assertThat(wrapper.item(1), `is`(equalTo(TestClass(2))))
        assertThat(wrapper.item(2), `is`(equalTo(TestClass(3))))
    }

    @Test
    fun replace_item() {
        wrapper.replaceData(listOf(TestClass(1), TestClass(2), TestClass(3)))
        wrapper.replaceItem(1, TestClass(4))
        assertThat(wrapper.item(0), `is`(equalTo(TestClass(1))))
        assertThat(wrapper.item(1), `is`(equalTo(TestClass(4))))
        assertThat(wrapper.item(2), `is`(equalTo(TestClass(3))))
    }

    @Test
    fun replace_data() {
        wrapper.replaceData(listOf(TestClass(1), TestClass(2), TestClass(3)))
        wrapper.replaceData(listOf(TestClass(4), TestClass(5), TestClass(6)))
        assertThat(wrapper.item(0), `is`(equalTo(TestClass(4))))
        assertThat(wrapper.item(1), `is`(equalTo(TestClass(5))))
        assertThat(wrapper.item(2), `is`(equalTo(TestClass(6))))
    }

    @Test
    fun modify_data() {
        wrapper.replaceData(listOf(TestClass(1), TestClass(2), TestClass(3)))
        wrapper.modifyData { input -> TestClass(input!!.id + 3) }
        assertThat(wrapper.item(0), `is`(equalTo(TestClass(4))))
        assertThat(wrapper.item(1), `is`(equalTo(TestClass(5))))
        assertThat(wrapper.item(2), `is`(equalTo(TestClass(6))))
    }

    private data class TestClass(val id: Int)
}